### PR TITLE
Ll 477

### LINKED ISSUE #477: Add Search by Streamgage Id functionality
- [x] Create a search box that takes in a single or multiple streamgage ids
- [x] Compare the ids to the site_no column and filter/zoom accordingly

### LINKED ISSUE #478: Add filter for gage types
There are 4 columns in the streamflow_gages dataset where a 1 indicates the gage is part of that subset and a 0 indicates it is not: nldi, swim, gfv1d1, and camels.

- [x] Create a selector to filter gages based on these columns

### PR DESCRIPTION
**Pull request by lluddenusgs**
# Pull Request Code Review

## Overview
To add functionality to the model_eviz app by, adding a slidebar, and search functionality to the model via the site_no
## Links
- *Issues*: Close #477 Closes #478
- *Inputs*: User's desired settings for the app (such as map, particular points)
- *Outputs*: User's desired output based on input

## Summary of Changes

- Sidebar added, 478 logic finished, 477 finished

## How to Run
Standard mamba panels app setup

## Peer Reviewer Expectations

Expect workaround logic, not finalized logic flows, yet basic functionality met. 
## Not expected in Review

N/A

### CODE REVIEW

File: apps/model_eval_viz/app.py [hunk 1]
```diff
@@ -83,58 +84,64 @@ def _get_data(_filepath:str)->gpd.GeoDataFrame:
...
 
 # Widget setup to select multiple states
 state_list = list(states['shapeName'].unique())
 #sort alphabetically
 state_list.sort()
+
+streamgage_input = pn.widgets.TextInput(
+    name='Streamgage Site ID', 
+    placeholder='Streamgage Site ID #',
+
```
Comment (alaws-USGS) [inline apps/model_eval_viz/app.py:98]: Add a description such as this
`description="Enter a column delimited list e.g. 01022500, 01022502"`

File: apps/model_eval_viz/app.py [hunk 2]
```diff
@@ -83,58 +84,64 @@ def _get_data(_filepath:str)->gpd.GeoDataFrame:
...
 )
 
-
 base_map_options = {
     'OpenStreetMap': gv.tile_sources.OSM,
     'ESRI Imagery': gv.tile_sources.EsriImagery,
     'ESRI World Street Map': gv.tile_sources.EsriWorldStreetMap,
 }
 
+subset = ['nldi','swim','gfv1d1','camels']
```
Comment (alaws-USGS) [inline apps/model_eval_viz/app.py:113]: Let's add this constant to the config.py and called it something like STREAMGAGE_SUBSET

File: apps/model_eval_viz/app.py [hunk 3]
```diff
@@ -159,9 +166,38 @@ def display_states(state_list:list=state_selector.value)->gv.Polygons:
...
+    Returns: None
+    '''
+    sg_value = streamgage_input.value 
+    if re.search('[^0-9, ]',sg_value):
+        streamgage_input.value = '' 
+        entered_points.value = ''
+        streamgage_input.placeholder ="Expected Format:[id_no], [id_no2]..."
+    else:
+        entered_points.value = sg_value
+enter_id = pn.panel(pn.widgets.Button(name='Enter', button_type='primary'))
```
Comment (alaws-USGS) [inline apps/model_eval_viz/app.py:194]: This works and that is a great first step. Great work!

**Review by alaws-USGS — CHANGES_REQUESTED** [review]

File: apps/model_eval_viz/app.py [hunk 4]
```diff
@@ -26,22 +27,22 @@
...
 mapproj = ccrs.Mercator(central_longitude=0.0, min_latitude=-80.0, max_latitude=84.0, globe=None, latitude_true_scale=0.0)
 
-
-
 # Initialize setup for below functions
 hv.extension('bokeh')
-path = "./data/streamflow_gages_v1_n5390.csv"
 pn.extension("plotly", "vega")
 
 # Create a map template including rough borders for a start
```
[the remark this answers is not in the document]
Comment (lluddenusgs) [reply apps/model_eval_viz/app.py:48]: Deleted lines 49-52 as discussed

[the remark this answers is not in the document]
Comment (alaws-USGS) [reply apps/model_eval_viz/app.py:48]: `gv_us`

**Review by alaws-USGS — CHANGES_REQUESTED** [review]

File: apps/model_eval_viz/app.py [hunk 5]
```diff
@@ -159,9 +157,38 @@ def display_states(state_list:list=state_selector.value)->gv.Polygons:
...
+    else:
+        entered_points.value = sg_value
+enter_id = pn.panel(pn.widgets.Button(name='Enter', button_type='primary'))
+
+enter_id.on_click(enter_event)
+
+
+
+
+def display_points(state_list:list=state_selector.value,ids:str=entered_points.value, data_set:str=subset_selector.value)->gv.Points:
```
Comment (alaws-USGS) [inline apps/model_eval_viz/app.py:191]: I beleive you can delete all of the default values here as these will be passed by the pn.rx() function. 

You can then update data_set = data_set, as that will be passed by pn.rx().

File: apps/model_eval_viz/app.py [hunk 6]
```diff
@@ -109,32 +108,31 @@ def _get_data(_filepath:str)->gpd.GeoDataFrame:
...
-        map = base_map_options[map_selector.value]
-    else:
-        #TODO find initial extent states.geometry.total_bounds
-        map = gv.tile_sources.OSM
-    return map
-
+    basemap = base_map_options[map]
+    return basemap
+    
 # create a DynamicMap to allow Panel to link map_selector with a Geoviews(Holoviews under the hood) object
```
Comment (alaws-USGS) [inline apps/model_eval_viz/app.py:134]: update comment to use pn.rx() instead of DynamicMap

File: apps/model_eval_viz/app.py [hunk 7]
```diff
@@ -159,9 +157,38 @@ def display_states(state_list:list=state_selector.value)->gv.Polygons:
     
 # create a DynamicMap to allow Panel to link state_selector with a Geoviews(Holoviews under the hood) object
```
Comment (alaws-USGS) [inline apps/model_eval_viz/app.py:158]: update comment to use pn.rx() instead of DynamicMap

File: apps/model_eval_viz/app.py [hunk 8]
```diff
@@ -171,22 +198,32 @@ def display_points(state_list:list=state_selector.value)->gv.Points:
...
+            filt_points = filt_points[filt_points['site_no'].isin(id_list)]
+    if data_set != "":
+        filt_points = filt_points[(filt_points[data_set]==1).all(axis=1)] 
+    selected_points = gv.Points(filt_points).opts(**plot_opts,color='lightgreen', size=5)
 
-    return displayed_points
+
+    return selected_points
 
 # create a DynamicMap to allow Panel to link state_selector with a Geoviews(Holoviews under the hood) object
```
Comment (alaws-USGS) [inline apps/model_eval_viz/app.py:221]: update comment to use pn.rx() instead of DynamicMap

**Review by alaws-USGS — CHANGES_REQUESTED** [review]

**Comment by alaws-USGS** [comment]
@lluddenusgs Delete lines 68-69 as we are no longer using `gv_us`

**Review by alaws-USGS — APPROVED** [review]

**Pull request merged**